### PR TITLE
fix: Rewrite implementation and testing of shift intrinsics

### LINKED ISSUE #393: Input data in `test_mm_srli_epi32` seldom hit the values used in real world usage
`test_mm_srli_epi32` and other tests implemented in the similar way are testing with rare input data. Take `test_mm_srli_epi32` for example, the input value of `imm8` should last 32 in the most cases in real world. However, in the current implementation, the `imm8` we generated are always out of this scope. We should refactor it to another way which can test the functions properly

### PR DESCRIPTION
**Pull request by marktwtn**
Implementation:
Rewrite conditional statement with bitwise AND operation.
Fix _mm_slli_si128() and _mm_srli_si128() intrinsics
and rewrite as functions.
Remove extra conditional statements for simplicity.

Testing:
Rewrite conditional statement with bitwise AND operation.
Test shift value in a reasonable range.
Test shift intrinsics required immediate shift value with different
input data.
Make the testing code follow the sequence order as variables declaration,
C implementation, SIMD implmentation and validation.

Close #393.

### CODE REVIEW

File: sse2neon.h [hunk 1]
```diff
@@ -5390,31 +5390,21 @@ FORCE_INLINE __m128i _mm_sll_epi64(__m128i a, __m128i count)
 //   r7 := a7 << count
 //
 // https://msdn.microsoft.com/en-us/library/es73bcsy(v=vs.90).aspx
```
Comment (jserv) [inline sse2neon.h:5392]: Change to Intel Intrinsics website hyperlink.

File: sse2neon.h [hunk 2]
```diff
@@ -5439,21 +5426,16 @@ FORCE_INLINE __m128i _mm_slli_epi64(__m128i a, int imm)
...
-            ret = vreinterpretq_m128i_s8(vextq_s8(                      \
-                vdupq_n_s8(0), vreinterpretq_s8_m128i(a), 16 - (imm))); \
-        }                                                               \
-        ret;                                                            \
-    })
+FORCE_INLINE __m128i _mm_slli_si128(__m128i a, int imm)
+{
+    if (_sse2neon_unlikely(imm & ~15))
+        return _mm_setzero_si128();
+    uint8x16_t tmp[2];
```
Comment (jserv) [inline sse2neon.h:5433]: Rewrite as
```c
 uint8x16_t tmp[2] = { vdupq_n_u8(0), vreinterpretq_u8_m128i(a) };
```

File: sse2neon.h [hunk 3]
```diff
@@ -5697,21 +5673,15 @@ FORCE_INLINE __m128i _mm_srl_epi64(__m128i a, __m128i count)
 //   r := srl(a, imm*8)
 //
 // https://msdn.microsoft.com/en-us/library/305w28yz(v=vs.100).aspx
```
Comment (jserv) [inline sse2neon.h:5675]: Ditto. Change the hyperlink.

File: sse2neon.h (hunk 2 above)
Comment (marktwtn) [reply sse2neon.h]: Done.

File: sse2neon.h (hunk 3 above)
Comment (marktwtn) [reply sse2neon.h]: Done.

File: sse2neon.h (hunk 1 above)
Comment (marktwtn) [reply sse2neon.h]: Done.

File: sse2neon.h [hunk 4]
```diff
@@ -5374,47 +5374,41 @@ FORCE_INLINE __m128i _mm_sll_epi32(__m128i a, __m128i count)
...
+{
+    if (_sse2neon_unlikely(imm & ~15))
+        return _mm_setzero_si128();
+    return vreinterpretq_m128i_s16(
+        vshlq_s16(vreinterpretq_s16_m128i(a), vdupq_n_s16(imm)));
+}
 
 // Shifts the 4 signed or unsigned 32-bit integers in a left by count bits while
 // shifting in zeros. :
 // https://msdn.microsoft.com/en-us/library/z2k3bbtb%28v=vs.90%29.aspx
```
Comment (jserv) [inline sse2neon.h:5407]: Ditto. Change URL.

File: sse2neon.h [hunk 5]
```diff
@@ -5439,21 +5430,14 @@ FORCE_INLINE __m128i _mm_slli_epi64(__m128i a, int imm)
 //   r := a << (imm * 8)
 //
 // https://msdn.microsoft.com/en-us/library/34d3k2kt(v=vs.100).aspx
```
Comment (jserv) [inline sse2neon.h:5432]: Ditto. Change URL.

File: sse2neon.h [hunk 6]
```diff
@@ -5494,7 +5478,7 @@ FORCE_INLINE __m128d _mm_sqrt_sd(__m128d a, __m128d b)
 FORCE_INLINE __m128i _mm_sra_epi16(__m128i a, __m128i count)
```
Comment (jserv) [inline sse2neon.h:5478]: Ditto. Change URL.

File: sse2neon.h [hunk 7]
```diff
@@ -5511,7 +5495,7 @@ FORCE_INLINE __m128i _mm_sra_epi16(__m128i a, __m128i count)
 FORCE_INLINE __m128i _mm_sra_epi32(__m128i a, __m128i count)
```
Comment (jserv) [inline sse2neon.h:5495]: Ditto. Change URL.

File: sse2neon.h [hunk 8]
```diff
@@ -5566,7 +5550,7 @@ FORCE_INLINE __m128i _mm_srai_epi16(__m128i a, int imm)
 FORCE_INLINE __m128i _mm_srl_epi16(__m128i a, __m128i count)
```
Comment (jserv) [inline sse2neon.h:5550]: Ditto. Change URL.

File: sse2neon.h [hunk 9]
```diff
@@ -5585,7 +5569,7 @@ FORCE_INLINE __m128i _mm_srl_epi16(__m128i a, __m128i count)
 FORCE_INLINE __m128i _mm_srl_epi32(__m128i a, __m128i count)
```
Comment (jserv) [inline sse2neon.h:5569]: Ditto. Change URL.

File: sse2neon.h [hunk 10]
```diff
@@ -5602,7 +5586,7 @@ FORCE_INLINE __m128i _mm_srl_epi32(__m128i a, __m128i count)
 FORCE_INLINE __m128i _mm_srl_epi64(__m128i a, __m128i count)
```
Comment (jserv) [inline sse2neon.h:5586]: Ditto. Change URL.

File: sse2neon.h [hunk 11]
```diff
@@ -5625,13 +5609,11 @@ FORCE_INLINE __m128i _mm_srl_epi64(__m128i a, __m128i count)
 #define _mm_srli_epi16(a, imm)                                             \
```
Comment (jserv) [inline sse2neon.h:5609]: Ditto. Change URL.

File: sse2neon.h [hunk 12]
```diff
@@ -5653,13 +5635,11 @@ FORCE_INLINE __m128i _mm_srl_epi64(__m128i a, __m128i count)
 #define _mm_srli_epi32(a, imm)                                             \
```
Comment (jserv) [inline sse2neon.h:5635]: Ditto. Change URL.

**Comment by marktwtn** [comment]
@jserv I will change the URL in this pull request and raise another pull request to change the others based on the issue #93.

**Comment by jserv** [comment]
> @jserv I will change the URL in this pull request and raise another pull request to change the others based on the issue #93.

Got it. It does not harm if we update the comments of modified functions/macros here.

**Pull request merged**